README.md
Benchmark Functions for CEC'2013 Special Session and Competition on Niching Methods for Multimodal Function Optimization
========================================================================================================================
*Organizers : Xiaodong Li, Andries Engelbrecht, and Michael G. Epitropakis*

*Version    : 1.1*

*Developers : Michael G. Epitropakis and Xiaodong Li*

Please refer to:
- http://goanna.cs.rmit.edu.au/~xiaodong/cec13-niching/
- http://goanna.cs.rmit.edu.au/~xiaodong/cec13-niching/competition/


If you are using any material from the current competition please cite the
following work:

X. Li, A. Engelbrecht, and M.G. Epitropakis, ``Benchmark Functions for
CEC'2013 Special Session and Competition on Niching Methods for Multimodal
Function Optimization'', Technical Report, Evolutionary Computation and
Machine Learning Group, RMIT University, Australia, 2013.

--------------------------------------------------------------------------------
- About:
--------------------------------------------------------------------------------

In this folder you can find all necessary source files for the benchmark suite
of the CEC'2013 Special Session and Competition on Niching Methods for
Multimodal Function Optimization. 

The Test suite for the competition is implemented in MATLAB, JAVA, C/C++ and python.

--------------------------------------------------------------------------------
- Documentation:
--------------------------------------------------------------------------------

For more information please refer to the Technical Report of the Special
Session/Competition

--------------------------------------------------------------------------------
- Installation instructions:
--------------------------------------------------------------------------------
Please unpack the archive and extract its contents in a folder:

  unzip FILENAME.zip
  cd FILENAME

In the FILENAME folder you will find the following folder structure:

--------------------------------------------------------------------------------
- Directory Structure:
--------------------------------------------------------------------------------

After unpacking the archive file, you should end up with the following
structure:
<pre>

  ./			The MAIN directory, created when unpacking
   |
   +-- matlab		Source code of the benchmark functions in MATLAB
   |   |
   |   +- figs		Figures of the benchmark functions (for validation)
   |   |
   |   +- data		Data files for the benchmark suite 
   |
   +-- c++		Source code of the benchmark functions in C/C++
   |   |
   |   +- plots		Figures of the benchmark function (for validation)
   |   |
   |   +- data		Data files for the benchmark suite 
   |
   +-- java 	Source code of the benchmark functions in JAVA
   |   |
   |   +- CEC2013	Project files
   |
   +-- python	Source code of the benchmark functions in python
       |
       +- CEC2013	Project files
       |
       +- data		Data files for the benchmark suite 
</pre>

--------------------------------------------------------------------------------
-- MATLAB: 
--------------------------------------------------------------------------------
The matlab folder contains: 
<pre>
+ niching_func.m
|-- 	The source code of the benchmark functions
|
+ demo_suite.m
|-- 	A demonstration file for using the competition source code 
|
+ count_goptima.m
|-- 	function for counting the number of global optima in a population 
|
+ plots.m
|-- 	MATLAB script for reproducing 1D/2D plots for the benchmark functions
|
+ data/
|-- 	Folder with data files
|
+ figs/
|-- 	Folder with the 1D/2D figures of the benchmark suite
|
+ get_fgoptima.m
|-- 	For each function, get the global's optima fitness value
|
+ get_no_goptima.m
|-- 	For each function, get the amount of known global optima 
|
+ get_rho.m
|-- 	For each function, get the rho value
|
+ get_lb.m
|-- 	For each function, get the lower bounds of the optimization box
|
+ get_ub.m
|-- 	For each function, get the upper bounds of the optimization box
</pre>

--------------------------------------------------------------------------------
-- C/C++:
--------------------------------------------------------------------------------
The c++ folder contains: 

<pre>
+ main.cpp:
|-- 	Examples on how to use each benchmark functions.
|
+ cec2013.h cec2013.cpp
|-- 	Interface and implementation of the CEC 2013 competition
|
+ cfunction.h cfunction.cpp
|-- 	The implementation of the benchmark function suite.
|
+ plots.cpp
|-- 	Source code to produce data files for plotting the benchmark functions
|
+ plots/
|-- 	It includes plotting data files, and the benchmark functions' figures. 
|
+ plots/plots.m
|-- 	MATLAB script for reproducing 1D/2D plots for the benchmark functions
|
+ Makefile: 
|-- 	A simple Makefile for easy compilation. 
|
+ data/
|-- 	Folder with data files
|
+ rand2.h rand2.c
|-- 	A simple random number generator 
|
+ randfile
|-- 	In Linux just a symbolic link on /dev/urandom 
|
+ sortidx.h
|-- 	Help header file for sorting
</pre>

--------------------------------------------------------------------------------
- Compilation 
--------------------------------------------------------------------------------
The Makefile provides the following three options:
make clean 	: Cleans all unnecessary files.  

make all 	: Compiles the main.cpp file. To see a demonstration of all
		  benchmark functions, execute ./a.out

make plots 	: Compiles the plots.cpp file. To produce the necessary data
		  files execute ./plots. It will produce 12 data files in 
		  the plots/ folder, one for each benchmark function. To 
		  produce the figures just run the plots.m MATLAB script.

--------------------------------------------------------------------------------
-- Java:
--------------------------------------------------------------------------------

The Java folder contains all necessary files to use the competition function
set. Please follow the demonstration file ExampleUsage.java on how to use the
source code. To reproduce the figures from the technical report please execute
the Plots.java file.

--------------------------------------------------------------------------------
-- python:
--------------------------------------------------------------------------------

The python folder contains all necessary files to use the competition function
set. Please follow the demonstration file demo.py on how to use the
source code. 

--------------------------------------------------------------------------------
-- ACKNOWLEDGEMENT: 
--------------------------------------------------------------------------------

- We really want to thank Dr. Jerry Swan for his fruitful comments and
  suggestions that helped to improve, the design and the implementation of the
  competition in the Java programming language. 
- We also thank Dr. Daniel Molina Cabrera, for finding and correcting a bug in
  the Java source code.
- We also thank Dr. Catalin Stoean, for finding and correcting a bug in the
  source code.
- We also thank Dr. Mike Preuss, for his valuable help and contributions on the
  competition.
- We really want to thank Eduardo Nobre Luis, and Fabricio Olivetti de Franca for 
  providing us a first implementation of the competition in python. This version 
  help us to redesign and re-implement the required infrastructure in python. 
- We also thank Francesco Fracchia, for finding and correcting the fitness
  value accuracy for shubert function

--------------------------------------------------------------------------------
-- LICENSE
--------------------------------------------------------------------------------
This folder should contain a file with the license statement (LICENSE.txt)
--------------------------------------------------------------------------------
Copyright 2013, Michael G. Epitropakis, Xiaodong Li, and Andries Engelbrecht. 
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met: 

1. Redistributions of source code must retain the above copyright notice, this
   list of conditions and the following disclaimer. 
2. Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution. 

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

The views and conclusions contained in the software and documentation are those
of the authors and should not be interpreted as representing official policies, 
either expressed or implied, of the FreeBSD Project.

(This is the so-called Simplified BSD License AKA FreeBSD License)


CHANGELOG.txt
-------------------------------------------------------------------------------
-- All programming languages
-------------------------------------------------------------------------------
-- 4 April 2016, 21:00PM
Increased accuracy for the fitness values of shubert, and six_hump_camel_back.
Many thanks to Francesco Fracchia for noting this to us
-------------------------------------------------------------------------------
-- C++ version
-------------------------------------------------------------------------------
11 March 2013, 9:31PM
Correct number of global optima for function 18 from 6 to 8. Thanks to Catalin
Stoean
-cec2013.cpp line 68: nopt_[17] = 6; nopt_[18] = 8; nopt_[19] = 8;

-------------------------
4 March 2013, 12:43PM
We just changed the line 40 in cec2013.h:
double nopt_[20]; with int nopt_[20];
-------------------------------------------------------------------------------
-- Java source code
-------------------------------------------------------------------------------
11 March 2013, 9:31PM
Correct number of global optima for function 18 from 6 to 8. Thanks to Catalin
Stoean
-CEC2013.java, lines 72-74,
    private static int [] nopt_ = new int[] {
        2, 5, 1, 4, 2, 18, 36, 81, 216, 12,
        6, 8, 6, 6, 8, 6, 8, 6, 8, 8};

-------------------------
7 March 2013, 10:00AM
Found a bug in CEC2013.java, fixed method howManyGlobalOptimaInPopulation
The bug was in the Map procedure used to sort the population by its fitness.
-------------------------------------------------------------------------------
-- MATLAB source code
-------------------------------------------------------------------------------
-- 4 April 2016, 21:00PM
Added get_dimension.m get_maxfes.m functions and updated the demo_suite.m script
-------------------------
-- 11 March 2013, 9:31PM
Correct number of global optima for function 18 from 6 to 8. Thanks to Catalin
Stoean
-get_no_goptima.m, line 10,
nopt = [2 5 1 4 2 18 36 81 216 12 6 8 6 6 8 6 8 6 8 8];
-------------------------------------------------------------------------------
-- python source code
-------------------------------------------------------------------------------
3 April 2016, 21:00PM
Initial implementation of the competition's infrastructure in python. Special 
thanks to Eduardo Nobre Luis, and Fabricio Olivetti de Franca for providing us 
a first implementation in python.


c++/cec2013.cpp
/******************************************************************************
 * Version: 1.0
 * Last modified on: 21 January, 2013 
 * Developers: Michael G. Epitropakis, Xiaodong Li.
 *      email: mge_(AT)_cs_(DOT)_stir_(DOT)_ac_(DOT)_uk 
 *           : xiaodong_(DOT)_li_(AT)_rmit_(DOT)_edu_(DOT)_au 
 * ***************************************************************************/
#include "cec2013.h"
#include "sortidx.h"

using namespace std;

CEC2013::CEC2013(const int &nfunc) :
	nfunc_(0), cfunc_(NULL)
{
	assert(nfunc>0 && nfunc<=20);
	nfunc_ = nfunc;
	init_vars_();

	if (nfunc_ == 11) {
		cfunc_ = new CF1(dimensions_[nfunc_-1]);
	} else if (nfunc_ == 12) {
		cfunc_ = new CF2(dimensions_[nfunc_-1]);
	} else if (nfunc_ == 13 || nfunc_ == 14 || nfunc_ == 16 || nfunc_ == 18 ) {
		cfunc_ = new CF3(dimensions_[nfunc_-1]);
	} else if (nfunc_ == 15 || nfunc_ == 17 || nfunc_ == 19 || nfunc_ == 20 ) {
		cfunc_ = new CF4(dimensions_[nfunc_-1]);
	}
}

CEC2013::~CEC2013()
{
	if (cfunc_) delete cfunc_;
}
void CEC2013::init_vars_()
{
	/* specific dimensions for the competition */
	dimensions_[0] = dimensions_[1] = dimensions_[2] = 1;
	dimensions_[3] = dimensions_[4] = dimensions_[5] = dimensions_[6] = 2;
	dimensions_[7] = dimensions_[8] = 3;
	dimensions_[9] = dimensions_[10] = dimensions_[11] = dimensions_[12] = 2;
	dimensions_[13] = dimensions_[14] = 3;
	dimensions_[15] = dimensions_[16] = 5;
	dimensions_[17] = dimensions_[18] = 10;
	dimensions_[19] = 20;
	fopt_[0]  = 200.0;		//1
	fopt_[1]  = 1.0;		//1
	fopt_[2]  = 1.0;		//1
	fopt_[3]  = 200.0;		//2
	fopt_[4]  = 1.031628453489877;	//2
	fopt_[5]  = 186.7309088310239;	//2
	fopt_[6]  = 1.0;		//2
	fopt_[7]  = 2709.093505572820;	//3
	fopt_[8]  = 1.0;		//3
	fopt_[9]  = -2.0;		//2
	for (int i=10; i<20; ++i) { fopt_[i] = 0;}
	rho_[0] = rho_[1] = rho_[2] = rho_[3] = 0.01;
	rho_[4] = rho_[5] = 0.5;
	rho_[6] = 0.2;
	rho_[7] = 0.5;
	rho_[8] = 0.2;
	rho_[9] = rho_[10] = rho_[11] = rho_[12] = rho_[13] = 0.01;
	rho_[14] = rho_[15] = rho_[16] = rho_[17] = rho_[18] = rho_[19] = 0.01;
	nopt_[0] = 2; nopt_[1] = 5; nopt_[2] = 1; nopt_[3] = 4; nopt_[4] = 2;
	nopt_[5] = 18; nopt_[6] = 36; nopt_[7] = 81; nopt_[8] = 216;
	nopt_[9] = 12; nopt_[10] = 6; nopt_[11] = 8; nopt_[12] = 6;
	nopt_[13] = 6; nopt_[14] = 8; nopt_[15] = 6; nopt_[16] = 8;
	nopt_[17] = 6; nopt_[18] = 8; nopt_[19] = 8;

	maxfes_[0] = maxfes_[1] = maxfes_[2] = maxfes_[3] = maxfes_[4] = 50000;
	maxfes_[5] = maxfes_[6] = maxfes_[9] = maxfes_[10] = maxfes_[11] = maxfes_[12] = 200000;
	maxfes_[7] = maxfes_[8] = maxfes_[13] = maxfes_[14] = maxfes_[15] = 400000;
	maxfes_[16] = maxfes_[17] = maxfes_[18] = maxfes_[19] = 400000;
}

double CEC2013::get_lbound(const int &n) const
{
	assert(n>=0 && n<dimensions_[nfunc_-1]);
	double result(0);
	if (nfunc_ == 1 || nfunc_ == 2 || nfunc_ == 3) {
		result = 0;
	} else if (nfunc_ == 4) {
		result = -6;
	} else if (nfunc_ == 5) {
		double tmp[2];
		tmp[0] = -1.9;
		tmp[1] = -1.1;
		result = tmp[n];
	} else if (nfunc_ == 6 || nfunc_ == 8) {
		result = -10;
	} else if (nfunc_ == 7 || nfunc_ == 9) {
		result = 0.25;
	} else if (nfunc_ == 10) {
		result = 0;
	} else if (nfunc_ > 10) {
		result = cfunc_->get_lbound(n);
	}
	return result;
}

double CEC2013::get_ubound(const int &n) const
{
	assert(n>=0 && n<dimensions_[nfunc_-1]);
	double result(0);
	if (nfunc_ == 1) {
		result = 30;
	} else if (nfunc_ == 2 || nfunc_ == 3) {
		result = 1;
	} else if (nfunc_ == 4) {
		result = 6;
	} else if (nfunc_ == 5) {
		double tmp[2];
		tmp[0] = 1.9;
		tmp[1] = 1.1;
		result = tmp[n];
	} else if (nfunc_ == 6 || nfunc_ == 8) {
		result = 10;
	} else if (nfunc_ == 7 || nfunc_ == 9) {
		result = 10;
	} else if (nfunc_ == 10) {
		result = 1;
	} else if (nfunc_ > 10) {
		result = cfunc_->get_ubound(n);
	}
	return result;
}

tFitness CEC2013::evaluate(const double *x)
{
	/* function number: */
	const int dim = dimensions_[nfunc_-1];
	double result(0);
	if (nfunc_ == 1) {
		result =  five_uneven_peak_trap(x, dim);
	} else if (nfunc_ == 2) {
		result =  equal_maxima(x, dim);
	} else if (nfunc_ == 3) {
		result =  uneven_decreasing_maxima(x, dim);
	} else if (nfunc_ == 4) {
		result =  himmelblau(x, dim);
	} else if (nfunc_ == 5) {
		result =  six_hump_camel_back(x, dim);
	} else if (nfunc_ == 6 || nfunc_ == 8) {
		result =  shubert(x, dim);
	} else if (nfunc_ == 7 || nfunc_ == 9) {
		result =  vincent(x, dim);
	} else if (nfunc_ == 10) {
		result =  modified_rastrigin_all(x, dim);
	} else if (nfunc_ > 10) {
		result = cfunc_->evaluate(x);
	}
	return result;
}

tFitness CEC2013::evaluate(std::vector<double> x)
{
	const int dim = x.size();
	assert(dim == dimensions_[nfunc_-1]);
	double *tmpx = new double[dim];
	for (int i=0; i<dim; ++i) { tmpx[i] = x[i]; }

	double result = evaluate(tmpx);
	delete [] tmpx;

	return result;
}

int how_many_goptima(std::vector< std::vector<double> > pop, 
		std::vector< std::vector<double> > &seeds, CEC2013 *pFunc,
		const double &accuracy, const double &radius)
{
	std::vector<std::vector<double> > sorted_pop;
	/* Evaluate pop */
	vector<double> fits;//(pop.size());
	//int i=0;
	for (std::vector< std::vector<double> >::iterator it = pop.begin(); it != pop.end(); ++it) {
		fits.push_back( pFunc->evaluate(*it) );
		//cout << "fits: " << fits[i] << endl;
		//++i;
	}

	/* sorted_pop: sort population by fitness */
	vector<size_t> idx;
	sortIdx(idx, fits, DESCEND);
	for (size_t i=0; i< pop.size(); ++i) {
		sorted_pop.push_back( pop[ idx[i] ] );
	}

	/* find seeds in the temp population */
	get_seeds(sorted_pop, seeds, radius);

	/* Based on the accuracy: check which seeds are global optimizers */
	double count(0);
	for (std::vector< std::vector<double> >::iterator sit = seeds.begin(); sit != seeds.end(); ++sit) {
		/* evaluate seed */
		double seed_fitness = pFunc->evaluate(*sit);

		/* |F_seed - F_goptimum| <= accuracy */
		if ( fabs(seed_fitness - pFunc->get_fitness_goptima()) <= accuracy) {
			++count;
		}
		/* save time */
		if (count == pFunc->get_no_goptima()) break;
	}
	return count;
}

void print_pop(std::vector< std::vector<double> > pop)
{
	for (std::vector< std::vector<double> >::iterator it = pop.begin(); 
			it != pop.end(); ++it) {
		for (std::vector<double>::iterator jt = it->begin();
				jt != it->end(); ++jt) {
			cout << *jt << "\t";
		}
		cout << endl;
	}
}

/**
 * Input:  population vector of vectors,
 * Output: seeds
 */
void get_seeds(std::vector< std::vector<double> > &pop, std::vector< std::vector<double> > &seeds,
		const double &radius)
{
	/* remove any forgoten items in the seed list */
	seeds.clear();
	/* Determine the species seeds: 
	 * iterate through sorted population */
	for (std::vector< std::vector<double> >::iterator it = pop.begin(); it != pop.end(); ++it) {
		bool found(false);
		/* Iterate seeds */
		//cout << "-----" << endl;
		for (std::vector< std::vector<double> >::iterator sit = seeds.begin(); sit != seeds.end(); ++sit) {
			/* Calculate distance from seeds */
			double dist = get_eudist(*it, *sit);
			//cout << "DIST: " << dist << endl;
			/* If the Euclidean distance is less than the radius */
			if (dist <= radius) {
				found = true;
				break;
			}
		} //seeds

		/* If it is not similar to any other seed, then it is a new seed */
		if (!found) { seeds.push_back(*it); }
	} //pop
}

double get_eudist(std::vector<double> &v1, std::vector<double> &v2)
{
	assert(v1.size() == v2.size());
	//if (v1.size() != v2.size()) { exit(-1);}
	double res(0);
	for (size_t i=0; i < v1.size(); i++) {
		res += (v1[i] - v2[i]) * (v1[i] - v2[i]);
	}
	return sqrt(res);
}

